### PR TITLE
feature/add auth module

### PR DESCRIPTION
**Pull request by hunter-s70**

### CODE REVIEW

**Review by haltaction — COMMENTED** [review]
It's cool that you write own validation, but in usual cases you could use validation in DTO via decorators like `@IsEmail(), @IsString(), @MaxLength(16)`
https://docs.nestjs.com/techniques/validation#using-the-built-in-validationpipe

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      const user: User = await this.userService.findUser({id, email});
+      if (!user) {
+        errors.push('usr does not exist');
+      }
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async checkUserExistence(userDto: UserDto): Promise<string> {
```
Comment (haltaction) [inline src/auth/auth.service.ts]: If you need only email in parameters, better to accept only it in parameter - your function became more universal.

`checkUserExistence(email: string) {...}`
... `checkUserExistence(userDto.email)`

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    if (userError && !errors.length) {
+      errors.push(userError);
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async validateSignOutFields(token: string): Promise<string[] | null> {
+    const errors = [];
+    const tokenData: any = this.jwtService.decode(token);
```
Comment (haltaction) [inline src/auth/auth.service.ts:48]: it's not necessary to describe types in a local variable inside the function, especially if its _any_

File: src/auth/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    const {email, password} = userDto;
+    const user: User = await this.userService.findUser({email});
+    const isUserPass: boolean = await bcrypt.compare(password, user.password);
+
+    if (!user || !isUserPass) {
+      return 'invalid email or password';
+    }
+  }
+
+  checkUsername(username: string): string {
```
Comment (haltaction) [inline src/auth/auth.service.ts]: Those independent functions (only logic without dependencies) could extract to separate service

File: src/users/user.service.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+@Injectable()
+export class UserService {
+  constructor(
+    @InjectRepository(User)
+    private usersRepository: Repository<User>,
+  ) {}
+
+  async findAll(): Promise<User[]> {
+    return await this.usersRepository.find();
```
Comment (haltaction) [inline src/users/user.service.ts]: In `return await <Promise>` you could skip _await_

File: src/auth/auth.service.ts (hunk 2 above)
Comment (hunter-s70) [reply src/auth/auth.service.ts:48]: @haltaction there are errors without type:

Comment (haltaction) [reply src/auth/auth.service.ts:48]: It's probably the TSLint rule, usually, IDE allows to see the rule and change TSLint config. 
By the way, what is the full error message?

Comment (hunter-s70) [reply src/auth/auth.service.ts:48]: @haltaction 
- `property 'email' does not exist on type 'string' | {[key: string]: any}`
- `property 'id' does not exist on type 'string' | {[key: string]: any}`

**Pull request merged**